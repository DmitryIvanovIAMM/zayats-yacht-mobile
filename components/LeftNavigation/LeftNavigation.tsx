import { primary, secondary } from "@/constants/Colors";
import { getMenuLinks } from "@/helpers/menuLinks";
import { RelativePathString, useRouter } from "expo-router";
import { Divider, Menu } from "react-native-paper";

export interface LeftNavigationProps {
  setMenuIsOpen: (isOpen: boolean) => void;
  visible: boolean;
}

export const LeftNavigation = ({
  setMenuIsOpen,
  visible
}: LeftNavigationProps) => {
  const router = useRouter();
  const menuLinks = getMenuLinks(false);

  const handleMenuItemPress = (link: string = "/", section: string = "") => {
    setMenuIsOpen(false);
    if (!link) return;

    const routeLink = link.startsWith("/") ? link : `/${link}`;
    const routeSection = section
      ? routeLink + `?section=${section}`
      : routeLink;
    router.push(routeSection as RelativePathString);
  };

  return (
    <Menu
      visible={visible}
      onDismiss={() => setMenuIsOpen(false)}
      //anchor={<Button onPress={() => setMenuIsOpen(true)}>Show menu</Button>}
      anchor={{ x: -10, y: 110 }}
      contentStyle={{ backgroundColor: secondary.dark }}
    >
      <>
        {menuLinks.map((menuLinkItem, index) => {
          const { label, link, section } = menuLinkItem;
          return (
            <Menu.Item
              key={`menu-item-${index}`}
              onPress={() => handleMenuItemPress(link, section)}
              title={label}
              titleStyle={{ color: primary.contrastText }}
            />
          );
        })}
        <Divider />
      </>
    </Menu>
  );
};
