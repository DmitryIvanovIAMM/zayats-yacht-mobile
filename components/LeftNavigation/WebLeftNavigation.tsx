import { primary, secondary } from "@/constants/Colors";
import { getMenuLinks } from "@/helpers/menuLinks";
import { RelativePathString, useRouter } from "expo-router";
import React from "react";
import { Modal, StyleSheet, Text, TouchableOpacity } from "react-native";

export interface WebLeftNavigationProps {
  setMenuIsOpen: (isOpen: boolean) => void;
  visible: boolean;
}

export const WebLeftNavigation = ({
  setMenuIsOpen,
  visible
}: WebLeftNavigationProps) => {
  const router = useRouter();
  const menuLinks = getMenuLinks(false);

  const handleMenuItemPress = (link: string = "/", section: string = "") => {
    setMenuIsOpen(false);
    if (!link) return;

    const routeLink = link.startsWith("/") ? link : `/${link}`;
    const routeSection = section
      ? routeLink + `?section=${section}`
      : routeLink;

    router.push(routeSection as RelativePathString);
  };

  return (
    <Modal
      visible={visible}
      transparent={true}
      animationType="fade"
      onRequestClose={() => setMenuIsOpen(false)}
    >
      <TouchableOpacity
        style={styles.overlay}
        activeOpacity={1}
        onPress={() => setMenuIsOpen(false)}
      >
        <TouchableOpacity
          style={styles.menuContainer}
          activeOpacity={1}
          onPress={(e) => e.stopPropagation()}
        >
          {menuLinks.map((menuLinkItem, index) => {
            const { label, link, section } = menuLinkItem;
            return (
              <TouchableOpacity
                key={`menu-item-${index}`}
                onPress={() => handleMenuItemPress(link, section)}
                style={styles.menuItem}
              >
                <Text style={styles.menuItemText}>{label}</Text>
              </TouchableOpacity>
            );
          })}
        </TouchableOpacity>
      </TouchableOpacity>
    </Modal>
  );
};

const styles = StyleSheet.create({
  overlay: {
    flex: 1,
    backgroundColor: "transparent",
    justifyContent: "flex-start",
    alignItems: "flex-start",
    paddingTop: 60,
    paddingLeft: 10
  },
  menuContainer: {
    backgroundColor: secondary.dark,
    borderRadius: 4,
    minWidth: 200,
    shadowColor: "#000",
    shadowOffset: {
      width: 0,
      height: 2
    },
    shadowOpacity: 0.25,
    shadowRadius: 3.84,
    elevation: 5
  },
  menuItem: {
    paddingVertical: 12,
    paddingHorizontal: 16,
    borderBottomWidth: 1,
    borderBottomColor: "rgba(255,255,255,0.1)"
  },
  menuItemText: {
    color: primary.contrastText,
    fontSize: 16
  }
});
