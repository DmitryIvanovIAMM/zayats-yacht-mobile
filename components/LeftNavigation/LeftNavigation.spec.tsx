import { fireEvent, render } from "@testing-library/react-native";
import React from "react";
import { LeftNavigation, LeftNavigationProps } from "./LeftNavigation";

// Mock expo-router
jest.mock("expo-router", () => {
  const mockPush = jest.fn();
  return {
    __esModule: true,
    useRouter: () => ({ push: mockPush }),
    mockPush,
  };
});

// Mock react-native-paper Menu and Divider
jest.mock("react-native-paper", () => {
  const React = require("react");
  const { View, TouchableOpacity, Text } = require("react-native");

  const Menu = ({
    children,
    visible,
    onDismiss,
    contentStyle,
    anchor,
  }: any) => {
    if (!visible) return null;

    return React.createElement(
      View,
      {
        testID: "menu",
        style: contentStyle,
        onPress: onDismiss,
      },
      children
    );
  };

  Menu.Item = ({ onPress, title, titleStyle, ...props }: any) =>
    React.createElement(
      TouchableOpacity,
      {
        testID: `menu-item-${title.toLowerCase().replace(/\s+/g, "-")}`,
        onPress,
        ...props,
      },
      React.createElement(Text, { style: titleStyle }, title)
    );

  const Divider = () => React.createElement(View, { testID: "menu-divider" });

  return { Menu, Divider };
});

// Mock menuLinks helper
jest.mock("@/helpers/menuLinks", () => ({
  getMenuLinks: jest.fn(() => [
    {
      label: "Schedule",
      link: "/",
      section: "schedule",
    },
    {
      label: "Get a Quote",
      link: "/quote-request",
    },
    {
      label: "Gallery",
      link: "/gallery",
    },
    {
      label: "Services",
      link: "/services",
    },
    {
      label: "Instructions",
      link: "/instructions",
    },
    {
      label: "Testimonials",
      link: "/",
      section: "testimonials",
    },
    {
      label: "About Us",
      link: "/",
      section: "about-us",
    },
    {
      label: "Contact Us",
      link: "/",
      section: "contact-us",
    },
  ]),
}));

describe("LeftNavigation", () => {
  const mockSetMenuIsOpen = jest.fn();
  const defaultProps: LeftNavigationProps = {
    setMenuIsOpen: mockSetMenuIsOpen,
    visible: true,
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("matches snapshot when visible", () => {
    const { toJSON } = render(<LeftNavigation {...defaultProps} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it("matches snapshot when not visible", () => {
    const { toJSON } = render(
      <LeftNavigation {...defaultProps} visible={false} />
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it("renders menu when visible is true", () => {
    const { getByTestId } = render(<LeftNavigation {...defaultProps} />);
    expect(getByTestId("menu")).toBeTruthy();
  });

  it("does not render menu when visible is false", () => {
    const { queryByTestId } = render(
      <LeftNavigation {...defaultProps} visible={false} />
    );
    expect(queryByTestId("menu")).toBeNull();
  });

  it("renders all menu items", () => {
    const { getByTestId } = render(<LeftNavigation {...defaultProps} />);

    expect(getByTestId("menu-item-schedule")).toBeTruthy();
    expect(getByTestId("menu-item-get-a-quote")).toBeTruthy();
    expect(getByTestId("menu-item-gallery")).toBeTruthy();
    expect(getByTestId("menu-item-services")).toBeTruthy();
    expect(getByTestId("menu-item-instructions")).toBeTruthy();
    expect(getByTestId("menu-item-testimonials")).toBeTruthy();
    expect(getByTestId("menu-item-about-us")).toBeTruthy();
    expect(getByTestId("menu-item-contact-us")).toBeTruthy();
  });

  it("renders divider at the end", () => {
    const { getByTestId } = render(<LeftNavigation {...defaultProps} />);
    expect(getByTestId("menu-divider")).toBeTruthy();
  });

  it("calls setMenuIsOpen with false when menu item is pressed", () => {
    const { getByTestId } = render(<LeftNavigation {...defaultProps} />);

    fireEvent.press(getByTestId("menu-item-schedule"));

    expect(mockSetMenuIsOpen).toHaveBeenCalledWith(false);
  });

  it("navigates to correct route when menu item is pressed", () => {
    const { getByTestId } = render(<LeftNavigation {...defaultProps} />);
    const { mockPush } = require("expo-router");

    fireEvent.press(getByTestId("menu-item-get-a-quote"));

    expect(mockPush).toHaveBeenCalledWith("/quote-request");
  });

  it("navigates to route with section when menu item has section", () => {
    const { getByTestId } = render(<LeftNavigation {...defaultProps} />);
    const { mockPush } = require("expo-router");

    fireEvent.press(getByTestId("menu-item-schedule"));

    expect(mockPush).toHaveBeenCalledWith("/?section=schedule");
  });

  it("handles empty link gracefully", () => {
    const { getByTestId } = render(<LeftNavigation {...defaultProps} />);
    const { mockPush } = require("expo-router");

    // Mock a menu item with empty link
    const { getMenuLinks } = require("@/helpers/menuLinks");
    getMenuLinks.mockReturnValueOnce([
      { label: "Empty Link", link: "", section: "" },
    ]);

    const { getByTestId: getByTestIdEmpty } = render(
      <LeftNavigation {...defaultProps} />
    );

    fireEvent.press(getByTestIdEmpty("menu-item-empty-link"));

    expect(mockPush).not.toHaveBeenCalled();
  });

  it("adds leading slash to link if not present", () => {
    const { getByTestId } = render(<LeftNavigation {...defaultProps} />);
    const { mockPush } = require("expo-router");

    // Mock a menu item without leading slash
    const { getMenuLinks } = require("@/helpers/menuLinks");
    getMenuLinks.mockReturnValueOnce([
      { label: "No Slash", link: "test-route", section: "" },
    ]);

    const { getByTestId: getByTestIdNoSlash } = render(
      <LeftNavigation {...defaultProps} />
    );

    fireEvent.press(getByTestIdNoSlash("menu-item-no-slash"));

    expect(mockPush).toHaveBeenCalledWith("/test-route");
  });

  it("calls setMenuIsOpen with false when menu is dismissed", () => {
    const { getByTestId } = render(<LeftNavigation {...defaultProps} />);

    fireEvent.press(getByTestId("menu"));

    expect(mockSetMenuIsOpen).toHaveBeenCalledWith(false);
  });

  it("logs navigation message to console", () => {
    const consoleSpy = jest.spyOn(console, "log").mockImplementation();
    const { getByTestId } = render(<LeftNavigation {...defaultProps} />);

    fireEvent.press(getByTestId("menu-item-schedule"));

    expect(consoleSpy).toHaveBeenCalledWith("Navigating to /");

    consoleSpy.mockRestore();
  });

  it("applies correct styling to menu", () => {
    const { getByTestId } = render(<LeftNavigation {...defaultProps} />);
    const menu = getByTestId("menu");

    expect(menu.props.style).toEqual({
      backgroundColor: "#006666", // secondary.dark
    });
  });

  it("applies correct styling to menu items", () => {
    const { getByTestId } = render(<LeftNavigation {...defaultProps} />);
    const menuItem = getByTestId("menu-item-schedule");
    const textElement = menuItem.children[0] as any;

    expect(textElement.props.style).toEqual({
      color: "#fff", // primary.contrastText
    });
  });
});
