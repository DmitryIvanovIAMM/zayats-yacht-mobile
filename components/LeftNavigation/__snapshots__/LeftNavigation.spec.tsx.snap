// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeftNavigation matches snapshot when not visible 1`] = `null`;

exports[`LeftNavigation matches snapshot when visible 1`] = `
<View
  onPress={[Function]}
  style={
    {
      "backgroundColor": "#006666",
    }
  }
  testID="menu"
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="menu-item-schedule"
  >
    <Text
      style={
        {
          "color": "#fff",
        }
      }
    >
      Schedule
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="menu-item-get-a-quote"
  >
    <Text
      style={
        {
          "color": "#fff",
        }
      }
    >
      Get a Quote
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="menu-item-gallery"
  >
    <Text
      style={
        {
          "color": "#fff",
        }
      }
    >
      Gallery
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="menu-item-services"
  >
    <Text
      style={
        {
          "color": "#fff",
        }
      }
    >
      Services
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="menu-item-instructions"
  >
    <Text
      style={
        {
          "color": "#fff",
        }
      }
    >
      Instructions
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="menu-item-testimonials"
  >
    <Text
      style={
        {
          "color": "#fff",
        }
      }
    >
      Testimonials
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="menu-item-about-us"
  >
    <Text
      style={
        {
          "color": "#fff",
        }
      }
    >
      About Us
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="menu-item-contact-us"
  >
    <Text
      style={
        {
          "color": "#fff",
        }
      }
    >
      Contact Us
    </Text>
  </View>
  <View
    testID="menu-divider"
  />
</View>
`;
